### PR TITLE
Preliminary changes to add aarch64 backend (round 2)

### PR DESCRIPTION
**Pull request by BreadFish64**
successor to #352 
I apologize for the glaring issues in that last PR, I was rushed for time that day and wanted early feedback.
Everything mentioned in the last PR should be resolved now 👍 
Any changes out of the backend area were to resolve compiler warnings or add things to common that dolphin's emitter needs.
A *much cleaner* dolphin diff can be found here:
https://github.com/BreadFish64/dolphin/commit/b5c7ff44679c0b28ccb515d66a2f56c477b46182

### CODE REVIEW

File: src/common/fp/op/FPRSqrtEstimate.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ static u8 RecipSqrtEstimate(u64 a) {
             }
 
             // Round to nearest u0.8 (with implied set integer bit).
-            result[i] = static_cast<u8>((b + 1) / 2);
+            result[static_cast<LUT::size_type>(i)] = static_cast<u8>((b + 1) / 2);
```
Comment (lioncash) [inline src/common/fp/op/FPRSqrtEstimate.cpp]: What does this resolve in particular?

File: src/backend/A64/dolphin_emitter/a64_emitter.cpp [hunk 2]
```diff
@@ -0,0 +1,3729 @@
...
+        *shift = true;
+        return true;
+    }
+    return false;
+}
+
+// For AND/TST/ORR/EOR etc
+bool IsImmLogical(uint64_t value, unsigned int width, unsigned int* n, unsigned int* imm_s,
+                  unsigned int* imm_r) {
+    // DCHECK((n != NULL) && (imm_s != NULL) && (imm_r != NULL));
```
Comment (lioncash) [inline src/backend/A64/dolphin_emitter/a64_emitter.cpp]: This is leftover commented out code.

File: src/backend/A64/dolphin_emitter/a64_emitter.cpp [hunk 3]
```diff
@@ -0,0 +1,3729 @@
...
+                   __func__, branch.type, distance);
+        inst = (0x5 << 26) | MaskImm26(distance);
+        break;
+    case 6: // BL (unconditional)
+        ASSERT_MSG(IsInRangeImm26(distance), "%s(%d): Received too large distance: %" PRIx64,
+                   __func__, branch.type, distance);
+        inst = (0x25 << 26) | MaskImm26(distance);
+        break;
+    }
+    *(u32*)branch.ptr = inst;
```
Comment (lioncash) [inline src/backend/A64/dolphin_emitter/a64_emitter.cpp]: This line in particular was already fixed upstream in Dolphin, given this particular cast invokes undefined behavior. It should be backported here as well.

File: src/common/fp/op/FPRSqrtEstimate.cpp (hunk 1 above)
Comment (lioncash) [reply src/common/fp/op/FPRSqrtEstimate.cpp]: This likely applies to all of the casts following this one, then

File: src/frontend/A64/translate/impl/simd_two_register_misc.cpp [hunk 4]
```diff
@@ -457,7 +457,7 @@ bool TranslatorVisitor::FNEG_2(bool Q, bool sz, Vec Vn, Vec Vd) {
 
     const size_t datasize = Q ? 128 : 64;
     const size_t esize = sz ? 64 : 32;
-    const size_t mask_value = esize == 64 ? 0x8000000000000000 : 0x8000000080000000;
+    const size_t mask_value = static_cast<size_t>(esize == 64 ? 0x8000000000000000 : 0x8000000080000000);
```
Comment (MerryMage) [inline src/frontend/A64/translate/impl/simd_two_register_misc.cpp]: Change the type of `mask_value` to `u64` instead.

File: src/frontend/A64/translate/impl/impl.cpp [hunk 5]
```diff
@@ -55,7 +55,7 @@ boost::optional<TranslatorVisitor::BitMasks> TranslatorVisitor::DecodeBitMasks(b
 
     size_t esize = static_cast<size_t>(1) << len;
     u64 welem = Common::Ones<u64>(S + 1);
-    u64 telem = Common::Ones<u64>(d + 1);
+    u64 telem = Common::Ones<u64>(static_cast<size_t>(d + 1));
```
Comment (lioncash) [inline src/frontend/A64/translate/impl/impl.cpp]: Does this cause warnings on GCC, or are these also likely unnecessary?

Comment (BreadFish64) [reply src/frontend/A64/translate/impl/impl.cpp]: looks like those can be reverted as well 👍

Comment (lioncash) [reply src/frontend/A64/translate/impl/impl.cpp]: Nice 👌

**Comment by BreadFish64** [comment]
comments addressed 👍
had to do some weird rebasing cause the folder move was merged early, so I hope nothing gets screwed up

**Pull request merged**